source/extensions/quic_listeners/quiche/envoy_quic_utils.cc

```diff
@@ -127,5 +127,65 @@ createConnectionSocket(Network::Address::InstanceConstSharedPtr& peer_addr,
   return connection_socket;
 }
 
+bssl::UniquePtr<X509> parseDERCertificate(const std::string& der_bytes,
+                                          std::string* error_details) {
+  const uint8_t* data;
+  const uint8_t* orig_data;
+  orig_data = data = reinterpret_cast<const uint8_t*>(der_bytes.data());
+  bssl::UniquePtr<X509> cert(d2i_X509(nullptr, &data, der_bytes.size()));
+  if (!cert.get()) {
+    *error_details = "d2i_X509";
+    return nullptr;
+  }
+  if (data < orig_data || static_cast<size_t>(data - orig_data) != der_bytes.size()) {
+    // Trailing garbage.
+    return nullptr;
+  }
+  return cert;
+}
+
+int deduceSignatureAlgorithmFromPublicKey(const EVP_PKEY* public_key, std::string* error_details) {
+  int sign_alg = 0;
+  const int pkey_id = EVP_PKEY_id(public_key);
+  switch (pkey_id) {
+  case EVP_PKEY_EC: {
+    // We only support P-256 ECDSA today.
+    const EC_KEY* ecdsa_public_key = EVP_PKEY_get0_EC_KEY(public_key);
+    // Since we checked the key type above, this should be valid.
+    ASSERT(ecdsa_public_key != nullptr);
+    const EC_GROUP* ecdsa_group = EC_KEY_get0_group(ecdsa_public_key);
+    if (ecdsa_group == nullptr || EC_GROUP_get_curve_name(ecdsa_group) != NID_X9_62_prime256v1) {
+      *error_details = "Invalid leaf cert, only P-256 ECDSA certificates are supported";
+    }
+    sign_alg = SSL_SIGN_ECDSA_SECP256R1_SHA256;
```

Comment:
naive question (both here and below in RSA case): how was SHA256 chosen? Is it required by the QUIC spec, just the best choice, or something else? I'm not asking you to change it, I just want to understand.

Reply:
SHA-256 comes from the Google QUIC Crypto protocol - that's the hash algorithm used with certificate signatures in it.

Reply:
Thanks for the clarification. Probably worth a comment in the code along those lines.